### PR TITLE
python310Packages.buildPythonPackage: introduce pyproject option

### PR DESCRIPTION
**Pull request by figsoda**
## Description of changes

#253154

adds the `pyproject` option that conflicts with `format`, where `pyproject = true` is equivalent to `format = "pyproject"`, and `pyproject = false` is equivalent to `format = "other"`, and 3 more commits to demonstrate the conversion of `format = "pyproject" | "setuptools" | "other"` to the `pyproject` option

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by pbsds** [comment]
The documentation [here](https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md?plain=1#L170-L175) needs some love

**Comment by figsoda** [comment]
> The documentation [here](https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md?plain=1&rgh-link-date=2023-09-09T08%3A15%3A26Z#L170-L175) needs some love

replaced all the `format`s in the documentation with `pyproject`

File: doc/languages-frameworks/python.section.md [hunk 1]
```diff
@@ -167,12 +167,15 @@ following are specific to `buildPythonPackage`:
...
-* `format ? "setuptools"`: Format of the source. Valid options are
-  `"setuptools"`, `"pyproject"`, `"flit"`, `"wheel"`, and `"other"`.
-  `"setuptools"` is for when the source has a `setup.py` and `setuptools` is
-  used to build a wheel, `flit`, in case `flit` should be used to build a wheel,
-  and `wheel` in case a wheel is provided. Use `other` when a custom
-  `buildPhase` and/or `installPhase` is needed.
+* `pyproject`: Whether the pyproject format should be used. When set to `true`,
+  `pypaBuildHook` will be used, and you can add the required build dependencies
+  from `build-system.requires` to `nativeBuildInputs`. Note that the pyproject
+  format falls back to using `setuptools`, so you can use `pyproject = true`
```
Comment (FRidh) [inline doc/languages-frameworks/python.section.md:173]: How? `setuptools` is not supposed to be available by default.

Comment (figsoda) [reply doc/languages-frameworks/python.section.md:173]: by that I mean `pypaBuildHook` will try to use `setuptools`, but it still needs to be added to `nativeBuildInputs`

**Comment by figsoda** [comment]
fixed merge conflicts

**Review by Mic92 — APPROVED** [review]

**Pull request merged**